# Jedlueng Django Learning 17 January 2023 

end 28 January 2023 
- I learn how to create app web app with django 
- fail to deploy
- learn how to use bootstrap
- Need to learn it better next time. 

## Install Django and environment 

Terminal 

#Make directory 

mkdir vidly 

#Go in vidly directory 

cd vidly 

#Install Django 
pipenv install django==2.1 

#or 
pip install django 

## Start Django Project 

In [2]:
%pip install django 

  Using cached Django-4.1.5-py3-none-any.whl (8.1 MB)
  Using cached asgiref-3.6.0-py3-none-any.whl (23 kB)
  Using cached sqlparse-0.4.3-py3-none-any.whl (42 kB)


In [3]:
import django

In [4]:
# Start a project 
!django-admin startproject vidly . 
#This will create files in admin 

<img src="Image/django-admin.png" alt="django-admin" />


In [ ]:
#manage.py == operation for the website 

#Image above is a package for our projects 

### Run a server 

In [11]:
!python3 manage.py runserver 

Watching for file changes with StatReloader
Performing system checks...

System check identified no issues (0 silenced).

You have 18 unapplied migration(s). Your project may not work properly until you apply the migrations for app(s): admin, auth, contenttypes, sessions.
Run 'python manage.py migrate' to apply them.
January 18, 2023 - 12:22:39
Django version 4.1.5, using settings 'vidly.settings'
Starting development server at http://127.0.0.1:8000/
Quit the server with CONTROL-C.
^C


# First Django Application 

<img src="Image/Django-project.png" alt="Django-project" />

Django project consist of many application but working under one banner 

<img src="Image/Project-2.png" alt="Django-project" />

## Create a new application 

In [12]:
!python3 manage.py startapp movies 
#New directory will be created call movies 
#This is a new apps 

<img src="Image/Move.png" alt="Django-project" />

## View function

<img src="Image/view.png" alt="view.png" />

view the app


<img src="Image/View-2.png" alt="Django-project" />

# Views 

In [ ]:
# Open view.py under movies application 

<img src="Image/view-3.png" alt="Django-project" />

create a function that represent the main page of the application. 

View is easily a way to mapp urls 

** Use request in view.py >> map url for the application (urls.py) > map url for application (Urls.py on another files)

<img src="Image/url-pics-1.png" alt="URL-pics-1" />

THen 

<img src="Image/url-pics-2.png" alt="URL-pics-2" />

Then 

<img src="Image/url-pics-3.png" alt="URL-pics-3" />

## Run server to test 

In [1]:
!python3 manage.py runserver

Watching for file changes with StatReloader
Performing system checks...

System check identified no issues (0 silenced).

You have 18 unapplied migration(s). Your project may not work properly until you apply the migrations for app(s): admin, auth, contenttypes, sessions.
Run 'python manage.py migrate' to apply them.
January 19, 2023 - 13:55:24
Django version 4.1.5, using settings 'vidly.settings'
Starting development server at http://127.0.0.1:8000/
Quit the server with CONTROL-C.
^C


# Models 

models > use to represent application data 

> Goes to application models.py

<img src="Image/url-pics-1.png" alt="URL-pics-1" />

# migration 

When you create models and application, it is not migrated yet 

bringing the database up to date 

sqlite 

Migration is the way to migrate data in the database to the models application in Django 

<img src="Migration.png" alt="Migration" />

In [1]:
#Run a migration 
!python3 manage.py makemigrations 

No changes detected


## Install movies 

main app > setting.py 

install_apps > add application Config (in apps.py of the app)

<img src="Image/apps.png" alt="apps" />

<img src="Image/movie-config-2.png" alt="Migration" />

In [6]:
# Make migration again 
!python3 manage.py makemigrations 

Migrations for 'movies':
  movies/migrations/0001_initial.py
    - Create model Genre
    - Create model Movie


In [7]:
!python3 manage.py runserver 
# 19 Migrations 

Watching for file changes with StatReloader
Performing system checks...

System check identified no issues (0 silenced).

You have 19 unapplied migration(s). Your project may not work properly until you apply the migrations for app(s): admin, auth, contenttypes, movies, sessions.
Run 'python manage.py migrate' to apply them.
January 20, 2023 - 07:27:53
Django version 4.1.5, using settings 'vidly.settings'
Starting development server at http://127.0.0.1:8000/
Quit the server with CONTROL-C.
^C


## All migrate 

In [8]:
!python3 manage.py migrate 
# Now migration are all migrated 
#Including the database 

Operations to perform:
  Apply all migrations: admin, auth, contenttypes, movies, sessions
Running migrations:
  Applying contenttypes.0001_initial... OK
  Applying auth.0001_initial... OK
  Applying admin.0001_initial... OK
  Applying admin.0002_logentry_remove_auto_add... OK
  Applying admin.0003_logentry_add_action_flag_choices... OK
  Applying contenttypes.0002_remove_content_type_name... OK
  Applying auth.0002_alter_permission_name_max_length... OK
  Applying auth.0003_alter_user_email_max_length... OK
  Applying auth.0004_alter_user_username_opts... OK
  Applying auth.0005_alter_user_last_login_null... OK
  Applying auth.0006_require_contenttypes_0002... OK
  Applying auth.0007_alter_validators_add_error_messages... OK
  Applying auth.0008_alter_user_username_max_length... OK
  Applying auth.0009_alter_user_last_name_max_length... OK
  Applying auth.0010_alter_group_name_max_length... OK
  Applying auth.0011_update_proxy_permissions... OK
  Applying auth.0012_alter_user_first_na

<img src="SQL-3.png" alt="SQL-3" />

# Changing Models 

Create a new datetime attribute 

App > models.py 

<img src="dt.png" alt="dt" />

In [9]:
!python3 manage.py makemigrations 

Migrations for 'movies':
  movies/migrations/0002_movie_date_created.py
    - Add field date_created to movie


all migration will be recorded in the migration folder in the application folder 

Do not ever delete migration!

## SQL migrate command 

In [10]:
!python3 manage.py sqlmigrate movies 
#show new sql database when migrate 

usage: manage.py sqlmigrate [-h] [--database DATABASE] [--backwards]
                            [--version] [-v {0,1,2,3}] [--settings SETTINGS]
                            [--pythonpath PYTHONPATH] [--traceback]
                            [--no-color] [--force-color] [--skip-checks]
                            app_label migration_name
manage.py sqlmigrate: error: the following arguments are required: migration_name


In [12]:
!python3 manage.py sqlmigrate movies 0001

BEGIN;
--
-- Create model Genre
--
CREATE TABLE "movies_genre" ("id" integer NOT NULL PRIMARY KEY AUTOINCREMENT, "name" varchar(255) NOT NULL);
--
-- Create model Movie
--
CREATE TABLE "movies_movie" ("id" integer NOT NULL PRIMARY KEY AUTOINCREMENT, "title" varchar(255) NOT NULL, "release_year" integer NOT NULL, "number_in_stock" integer NOT NULL, "daily_rate" real NOT NULL, "genre_id" bigint NOT NULL REFERENCES "movies_genre" ("id") DEFERRABLE INITIALLY DEFERRED);
CREATE INDEX "movies_movie_genre_id_2b557368" ON "movies_movie" ("genre_id");
COMMIT;


# Administration Panel 

for people to fill in the code 

In [14]:
!python3 manage.py makemigrations 

No changes detected


In [16]:
!python3 manage.py migrate 

Operations to perform:
  Apply all migrations: admin, auth, contenttypes, movies, sessions
Running migrations:
  Applying movies.0002_movie_date_created... OK


In [19]:
!python3 manage.py runserver

Watching for file changes with StatReloader
Performing system checks...

System check identified no issues (0 silenced).
January 20, 2023 - 10:03:24
Django version 4.1.5, using settings 'vidly.settings'
Starting development server at http://127.0.0.1:8000/
Quit the server with CONTROL-C.
Not Found: /
[20/Jan/2023 10:03:26] "GET / HTTP/1.1" 404 2165
Not Found: /favicon.ico
[20/Jan/2023 10:03:26] "GET /favicon.ico HTTP/1.1" 404 2216
[20/Jan/2023 10:03:53] "GET /admin HTTP/1.1" 301 0
[20/Jan/2023 10:03:53] "GET /admin/ HTTP/1.1" 302 0
[20/Jan/2023 10:03:53] "GET /admin/login/?next=/admin/ HTTP/1.1" 200 2218
[20/Jan/2023 10:03:53] "GET /static/admin/css/nav_sidebar.css HTTP/1.1" 200 2619
[20/Jan/2023 10:03:53] "GET /static/admin/css/dark_mode.css HTTP/1.1" 200 796
[20/Jan/2023 10:03:53] "GET /static/admin/css/login.css HTTP/1.1" 200 958
[20/Jan/2023 10:03:53] "GET /static/admin/css/base.css HTTP/1.1" 200 20344
[20/Jan/2023 10:03:53] "GET /static/admin/css/responsive.css HTTP/1.1" 200 18854

http://127.0.0.1:8000/ <<< This is server. >>

http://127.0.0.1:8000/admin 

<img src="admin.png" alt="admin" />

## Create a super user

In [1]:
!python3 manage.py createsuperuser 


Username (leave blank to use 'jedsadaluengaramsuk'): ^C

Operation cancelled.


<img src="Image/Superuser-1.png" alt="superuser" />

pass: jedsada1

GO to the URL (Server) > Login 

In [2]:
!python3 manage.py runserver

Watching for file changes with StatReloader
Performing system checks...

System check identified no issues (0 silenced).
January 21, 2023 - 05:13:37
Django version 4.1.5, using settings 'vidly.settings'
Starting development server at http://127.0.0.1:8000/
Quit the server with CONTROL-C.
[21/Jan/2023 05:13:54] "GET /admin/ HTTP/1.1" 302 0
[21/Jan/2023 05:13:54] "GET /admin/login/?next=/admin/ HTTP/1.1" 200 2218
[21/Jan/2023 05:13:54] "GET /static/admin/css/base.css HTTP/1.1" 304 0
[21/Jan/2023 05:13:54] "GET /static/admin/css/dark_mode.css HTTP/1.1" 304 0
[21/Jan/2023 05:13:54] "GET /static/admin/css/responsive.css HTTP/1.1" 304 0
[21/Jan/2023 05:13:54] "GET /static/admin/css/nav_sidebar.css HTTP/1.1" 304 0
[21/Jan/2023 05:13:54] "GET /static/admin/css/login.css HTTP/1.1" 304 0
[21/Jan/2023 05:13:54] "GET /static/admin/js/nav_sidebar.js HTTP/1.1" 304 0
[21/Jan/2023 05:13:54] "GET /static/admin/css/fonts.css HTTP/1.1" 304 0
[21/Jan/2023 05:13:55] "GET /static/admin/fonts/Roboto-Regular-

http://127.0.0.1:8000/admin

<img src="Image/login.png" alt="login-1" />

<img src="Image/login-2.png" alt="login-2" />

<img src="Image/managing-user.png" alt="managing-user" />

## Customize admin interface 

app folder > admin.py 

<img src="Image/admin-3.png" alt="admin-3" />

<img src="Image/admin-4.png" alt="admin-4" />

# Cutomize the admin

app folder > model.py

<img src="Image/admin-5.png" alt="admin-5" /> 

I add action gerne but it show Genre object, Therefore, I need to customize it 

app folder >> models.py

<img src="Image/admin-6.png" alt="admin-6" /> 

Now the actual name is here 

<img src="Image/admin-7.png" alt="admin-7" /> 

## Add column 

App folder > admin.py 

<img src="admin-8.png" alt="admin-8" /> 

now on the admin, it will be like this (id and name)

<img src="Image/admin-9.png" alt="admin-9" /> 

## Add object in the application 

<img src="Image/admin-11.png" alt="admin-11" /> 

<img src="Image/admin-12.png" alt="admin-12" /> 

Still a native object therefore, need to change it to string 

app folder > admin.py

now it will be like this 

<img src="Image/admin-14.png" alt="admin-14" /> 

<img src="Image/admin-15.png" alt="admin-15" /> 

# Database Abstraction API

models classes (Django) > interact with SQL (database )

Therefore you need to fill the right SQL statement 

app folder > views.py to configue the html

<img src="api-1.png" alt="api-1" /> 

import Movie object from model into views 

<img src="api-4.png" alt="api-4" /> 

then the return will be like this (SQL ready)

<img src="Image/api-2.png" alt="api-2" /> 

# Template 

return HTML content in the view function 

app folder > views.py 

first , need to import render function 

In [3]:
from django.shortcuts import render

<img src="Image/template-1.png" alt="template-1" /> 

Create folder called template in the application 

Then create HTML file as a template


<img src="Image/template-2.png" alt="template-2" /> 

<img src="Image/template-3.png" alt="template-3" /> 

<img src="Image/template-4.png" alt="template-4" /> 

now initialized html

## Problems with indexing (Django)

django try to find the html file by looping through all the install apps. Therefore, use relative path instead of actual file name is better 

<img src="Image/template-5.png" alt="template-5" /> 

<img src="Image/template-6.png" alt="template-6" /> 

# Bootstrap

Give style to our application 

https://getbootstrap.com/docs/5.3/getting-started/introduction/

https://getbootstrap.com/docs/5.0/getting-started/introduction/

Using the starter template > add base.html in the template file(application folder )

Change the title and add your content in body tag 

<img src="Image/bootstrap-1.png" alt="bootstrap-1" /> 

now the block will be connect to another html page 

go to index.html

using extend tag (django)

then use block tag to sync two html together 

<img src="Image/boot-4.png" alt="boot-4" /> 

<img src="Image/boot-5.png" alt="boot-5" /> 

now the bootstrap is activated 

# Customizing the Layout

Insert a navbar (code in the document) > use as a link one 

In [ ]:
<!-- As a link -->
<nav class="navbar navbar-light bg-light">
  <div class="container-fluid">
    <a class="navbar-brand" href="#">Navbar</a>
  </div>
</nav>

<!-- As a heading -->
<nav class="navbar navbar-light bg-light">
  <div class="container-fluid">
    <span class="navbar-brand mb-0 h1">Navbar</span>
  </div>
</nav>

<img src="Image/add-navbar.png" alt="add-navbar" /> 

add navbar to body 

<img src="Image/nav-1.png" alt="nav-1"/>

Now the navbar is there

## Container

<img src="Image/nav-2.png" alt="nav-2"/>

<img src="Image/nav-4.png" alt="nav-4"/>

## Customize table 

<img src="Image/nav-5.png" alt="nav-5"/>

add more classes now table will be customized 

# Sharing template accross multiple apps 

create new folder (at root ) called template 

then move base.html to universal template 

change the relative path of that index base.html to just a name (index.html)

<img src="Image/uni-1.png" alt="uni-1"/>

go to settings.py of the main app (vidly)

reconfigure the template settings 

add new template folder in DIRS (using os)

<img src=Image/uni-2.png" alt="uni-2"/>

now we can use base template everywhere 

# URL Parameters 

<img src="Image/url-1.png" alt="url-1"/>

Create URL for each items 

movies/1 = first movie 
moves/2 = second movie

configue in urls.py (movies app) 

<img src="Image/url-2.png" alt="url-2"/>

you also need to create view function (views.detail)

<img src="Image/url-3.png" alt="url-3"/>

http://127.0.0.1:8000/movies/1 will show 

<img src="Image/url-4.png" alt="url-4"/>

# Getting a Single Object

Make a page for items 

go to views.py (movies)

<img src="ob-1.png" alt="ob-1"/>

Then create deatil.html in the template folder 

<img src="ob-2.png" alt="ob-2"/>

<img src="ob-3.png" alt="ob-3"/>

# Raising 404 Errors 

return https 404 errors when the key does not exist 

Go to views.py ( of apps )

Create try and except cause

<img src="404-1.png" alt="404-1"/>

<img src="404-2.png" alt="404-2"/>

## Way 2 (Django)

views.py > import get object or 404 (getobjector404)

<img src="Image/404-3.png" alt="404-3"/>

#  Referencing Urls

<img src="Image/ul-1.png" alt="ul-1"/>

when you click name of the object > show movies details 

go to index.html
add link on movie title 

<img src="Image/ul-2.png" alt="ul-2"/>

now you can click here 

<img src="Image/ul-3.png" alt="ul-3"/>

## Second way >> using Django 

<img src="Image/ul-4.png" alt="ul-4"/>

result will be the same 

## Third way 

go to urls.py

<img src="Image/ul-5.png" alt="ul-5"/>

<img src="Image/ul-6.png" alt="ul-6"/>

result wil be the same 

# Creating APIs 

django tastypie 

In [1]:
%pip install djangorestframework 

     |████████████████████████████████| 1.1 MB 2.8 MB/s eta 0:00:01
Note: you may need to restart the kernel to use updated packages.


In [8]:
%pip install django-tastypie
 

Note: you may need to restart the kernel to use updated packages.


In [9]:
# Start api project 
!python3 manage.py startapp api

Now we need to register api at the install app > setting.py of the main app 

<img src="Image/apii-1.png" alt="api-1"/>

This is an automatically created application for apis 

## Create models for apis 

models.py (APIs)

<img src="Image/apii-3.png" alt="api-3"/>

then go to urls.py for the main app 

<img src="Image/apii-4.png" alt="api-4"/>

 http://127.0.0.1:8000/api/movies 

<img src="Image/apii-5.png" alt="api-5"/>

## Exclude in api 

go to models.py of apis 

<img src="Image/apii-6.png" alt="api-6"/>

<img src="Image/apii-7.png" alt="api-7"/>

now there's no date-created 

# Adding the homepage

go to urls.py of the main app

create views.py for the main app

<img src="Image/home-1.png" alt="home-1"/>

go to urls.py of the main app

<img src="Image/home-2.png" alt="home-2"/>

then add home.html to templates folder 

<img src="Image/home-4.png" alt="home-4"/>

test with html to get the right page 

<img src="Image/home-5.png" alt="home-5"/>

now customize home.html to include link

<img src="Image/home-6.png" alt="home-6"/>

<img src="Image/home-7.png" alt="home-7"/>

# Deployment Preparation 

Install git

## Intall Heroku

https://www.heroku.com/

<img src="Image/heroku-1.png" alt="home-7"/>

### Heroku CLI

https://devcenter.heroku.com/articles/heroku-cli

In [2]:
# Install home brew 

!/bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"

==> Checking for `sudo` access (which may request your password)...
Password:
sudo: a password is required
Need sudo access on macOS (e.g. the user jedsadaluengaramsuk needs to be an Administrator)!


In [9]:
!brew tap heroku/brew && brew install heroku


zsh:1: command not found: brew


## Install Gnicorn

In [2]:
!pipenv install gnicore


We recommend setting this in ~/.profile (or equivalent) for proper expected behavior.
Creating a virtualenv for this project...
Pipfile: /Users/jedsadaluengaramsuk/Documents/GitHub/Django-learning /django-/Pipfile
Using /usr/bin/python3 (3.9.6) to create virtualenv...
⠇ Creating virtual environment.....created virtual environment CPython3.9.6.final.0-64 in 648ms
  creator CPython3macOsFramework(dest=/Users/jedsadaluengaramsuk/.local/share/virtualenvs/django--LIRP7-CQ, clear=False, no_vcs_ignore=False, global=False)
  seeder FromAppData(download=False, pip=bundle, setuptools=bundle, wheel=bundle, via=copy, app_data_dir=/Users/jedsadaluengaramsuk/Library/Application Support/virtualenv)
    added seed packages: pip==22.3.1, setuptools==65.6.3, wheel==0.38.4
  activators BashActivator,CShellActivator,FishActivator,NushellActivator,PowerShellActivator,PythonActivator

⠏ Creating virtual environment...✔ Successfully created virtual environment!
⠏ Creating virtual environment...
Virtualenv lo

## Procfile 

add Procfile at root 

file for heroku

<img src="Image/heroku-2.png" alt="heroku-2"/>

vidly.wsgi = webserver deployement 

## Implementing Heroku 

### Creating static file

main app > settings.py 

<img src="Image/heroku-4.png" alt="heroku-4"/>

run collectstatic 

In [3]:
!python3 manage.py collectstatic


You have requested to collect static files at the destination
location as specified in your settings:

    /Users/jedsadaluengaramsuk/Documents/GitHub/Django-learning /django-/templates

This will overwrite existing files!
Are you sure you want to do this?

Type 'yes' to continue, or 'no' to cancel: ^C
Traceback (most recent call last):
  File "manage.py", line 22, in <module>
    main()
  File "manage.py", line 18, in main
    execute_from_command_line(sys.argv)
  File "/Users/jedsadaluengaramsuk/opt/anaconda3/lib/python3.8/site-packages/django/core/management/__init__.py", line 446, in execute_from_command_line
    utility.execute()
  File "/Users/jedsadaluengaramsuk/opt/anaconda3/lib/python3.8/site-packages/django/core/management/__init__.py", line 440, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/Users/jedsadaluengaramsuk/opt/anaconda3/lib/python3.8/site-packages/django/core/management/base.py", line 402, in run_from_argv
    self.execute(*args, 

now all files will be copy to the templates folder 

<img src="Image/heroku-6.png" alt="heroku-6"/>

## Install whitenoise

this is for heroku

In [4]:
!pipenv install whitenoise

Installing whitenoise...
⠋ ✔ Installation Succeeded..
Pipfile.lock not found, creating...
Locking [packages] dependencies...
⠇ ✔ Success!dependencies.....
Locking [dev-packages] dependencies...
Updated Pipfile.lock (40deb82533e1aacdb73fb0d6b9ea924523516ad5e69cbe0c6dc292f38c1af3f7)!
Installing dependencies from Pipfile.lock (1af3f7)...
To activate this project's virtualenv, run pipenv shell.
Alternatively, run a command inside the virtualenv with pipenv run.


http://whitenoise.evans.io/en/latest/

http://whitenoise.evans.io/en/latest/django.html

follow the steps 

add this in settings.py

<img src="Image/middlewear-1.png" alt="heroku-6"/>

# Deployement Process 

<img src="Image/deploy-1.png" alt="deploy-1"/>

push from local to Heroku git 

## Git test 

In [5]:
#Stage git 
!git init

Reinitialized existing Git repository in /Users/jedsadaluengaramsuk/Documents/GitHub/Django-learning /django-/.git/


In [6]:
!git add .

add the staging 

In [7]:
!git commit -m test 

[main 98e3bc1] test
 151 files changed, 30670 insertions(+), 2 deletions(-)
 create mode 100644 Image/deploy-1.png
 create mode 100644 Image/heroku-2.png
 create mode 100644 Image/heroku-4.png
 create mode 100644 Image/heroku-6.png
 create mode 100644 Image/middlewear-1.png
 create mode 100644 Pipfile
 create mode 100644 Pipfile.lock
 create mode 100644 Procfile
 create mode 100644 api/__pycache__/__init__.cpython-38.pyc
 create mode 100644 api/__pycache__/admin.cpython-38.pyc
 create mode 100644 api/__pycache__/apps.cpython-38.pyc
 create mode 100644 api/__pycache__/models.cpython-38.pyc
 create mode 100644 templates/admin/css/autocomplete.css
 create mode 100644 templates/admin/css/base.css
 create mode 100644 templates/admin/css/changelists.css
 create mode 100644 templates/admin/css/dark_mode.css
 create mode 100644 templates/admin/css/dashboard.css
 create mode 100644 templates/admin/css/fonts.css
 create mode 100644 templates/admin/css/forms.css
 create mode 100644 templates/admi

## Heroku push 

In [1]:
!heroku login 

zsh:1: command not found: heroku


** everytime you do you need to install brew and use heroku again 

Install brew https://stackoverflow.com/questions/66666134/how-to-install-homebrew-on-m1-mac

login to heroku 

In [ ]:
!heroku login 

Finished 

In [1]:
!heroku create 

zsh:1: command not found: heroku


<img src="Image/heroku-5.png" alt="heroku-5"/>

need to add card details 

<img src="Image/heroku-7.png" alt="heroku-7"/>

https://arcane-reaches-72002.herokuapp.com/ | https://git.heroku.com/arcane-reaches-72002.git

First link will be this 

<img src="Image/heroku-8.png" alt="heroku-8"/>

## Push files to heroku 

In [3]:
!git push heroku main

Enumerating objects: 410, done.
Counting objects: 100% (410/410), done.
Delta compression using up to 8 threads
Compressing objects: 100% (341/341), done.
Writing objects: 100% (410/410), 9.12 MiB | 1.30 MiB/s, done.
Total 410 (delta 119), reused 197 (delta 58), pack-reused 0
remote: Compressing source files... done.
remote: Building source:
remote: 
remote: -----> Building on the Heroku-22 stack
remote: -----> Determining which buildpack to use for this app
remote: -----> Python app detected
remote: -----> Using Python version specified in Pipfile.lock
remote: -----> Installing python-3.9.16
remote: -----> Installing pip 22.3.1, setuptools 63.4.3 and wheel 0.37.1
remote: -----> Installing dependencies with Pipenv 2020.11.15
remote:        Installing dependencies from Pipfile.lock (1af3f7)...
remote: -----> Installing SQLite3
remote: -----> Discovering process types
remote:        Procfile declares types -> web
remote: 
remote: -----> Compressing...
remote:        Done: 48.1M
remote: -

Tell heroku to use web-server 

https://devcenter.heroku.com/articles/scaling

In [4]:
#After configure heroku 
heroku ps:scale web=2

zsh:1: command not found: heroku


<img src="Image/heroku-10.png" alt="heroku-10"/>

In [ ]:
!heroku open 

<img src="Image/heroku-11.png" alt="heroku-11"/>

still need to fix bug 

open up settings.py 

go to ALLOWED_HOST 

<img src="Image/heroku-14.png" alt="heroku-14"/>

fixing and push to git 

In [ ]:
!git add . 

In [ ]:
!git commit -m 'fixing '

In [ ]:
!git push heroku main

add this to heroku (buildpack)

<img src="Image/heroku-20.png" alt="heroku-20"/>

https://elements.heroku.com/buildpacks/benoitc/gunicorn

# Summary 

OK I can do every step except deployment. 

You can look and follow turtorial later. Now I'm done. Thank you!